Trivial change arf

ezyang opened the pull request:
Trivial change arf

Summary: none

Differential Revision: https://phabricator.intern.facebook.com/D9514646

apaszke commented on torch/csrc/jit/assert.cpp:
```diff
@@ -15,9 +15,10 @@ barf(const char *fmt, ...)
+  // If you uncomment this, gdb will break at this point so you
+  // can inspect the backtrace.  DO NOT commit this, since this
+  // will cause a hard failure rather than a soft one.
+  // __builtin_trap();
```
Right. This is exactly what I said in the previous PR :D Just use `catch throw`

apaszke commented on torch/csrc/jit/ir.cpp:
```diff
@@ -61,4 +61,61 @@ std::ostream& operator<<(std::ostream & out, Graph & g) {
+    JIT_ASSERT(!known_inputs.count(n));
+    JIT_ASSERT(n->kind_ == NodeKind::Param);
+    known_inputs.insert(n);
+  }
```
I think a simpler way to write this is:
```cpp
// this is used multiple times so useful to define.
// use ordered set so we can use std::includes
using node_set = std::set<Node*>; 
// I don't like typing these...
#define ALL_OF(container) container.begin(), container.end()

node_set all_nodes_set(ALL_OF(all_nodes));
JIT_ASSERT(all_nodes_set.size() == all_nodes.size());

// These should be included in all_nodes
node_set nodes_set(ALL_OF(nodes_));
node_set inputs_set(ALL_OF(inputs_));
node_set output_set {output_};
JIT_ASSERT(std::includes(ALL_OF(all_nodes_set), ALL_OF(nodes_set));
JIT_ASSERT(std::includes(ALL_OF(all_nodes_set), ALL_OF(inputs_set));
JIT_ASSERT(std::includes(ALL_OF(all_nodes_set), ALL_OF(output_set));

// They should sum to all nodes
node_set sum_set;
sum_set.insert(ALL_OF(nodes_set));
sum_set.insert(ALL_OF(inputs_set));
sum_set.insert(ALL_OF(output_set));
JIT_ASSERT(std::equals(ALL_OF(all_nodes_set), ALL_OF(sum_set));

// And they should be disjoint
JIT_ASSERT(std::find_fist_of(ALL_OF(sum_set), ALL_OF(input_set)) == sum_set.end());
JIT_ASSERT(std::find_fist_of(ALL_OF(output_set), ALL_OF(input_set)) == output_set.end());
JIT_ASSERT(std::find_fist_of(ALL_OF(output_set), ALL_OF(sum_set)) == output_set.end());

// All inputs should be Params
JIT_ASSERT(std::all_of(ALL_OF(input_set), [](Node* n) { return n->kind_ == NodeKind::Param; }));
```

ezyang replied:
NB: nodes_set includes inputs and outputs, so there is a stronger "They should sum to all nodes" invariant. (This is why writing down the invariants as code is good :)

apaszke commented on torch/csrc/jit/ir.h:
```diff
@@ -276,12 +276,24 @@ TH_DISALLOW_COPY_AND_ASSIGN(Graph);
   const node_list & outputs() {
     return output_->inputs();
   }
+  const node_list & outputs() const {
```
I think you can remove non-const methods if they're the same. Const methods can be used on non-const objects

ezyang commented:
This needs rebasing.

ezyang committed:
Trivial change arf

Summary: none

Differential Revision: https://phabricator.intern.facebook.com/D9514646
